### ISSUE TITLE
Unit tests fail when using Java 8 to build maven package

### ISSUE THREAD

**Issue by marvkis** [jkuhnert/ognl#16]
Hi,

When doing a "mvn test" with Java 8 the tests fail with these error messages:

```
Results :

Failed tests:
  messages.format('ShowAllCount', one) (foo)(org.ognl.test.MethodTest)

Tests in error:
  {'1','2','3'}(org.ognl.test.ArrayElementsTest)
  { true, !false }(org.ognl.test.ArrayElementsTest)
  {stringValue, getMap()}(org.ognl.test.PropertyTest)
  {'stringValue', map["test"].map["size"]}(org.ognl.test.PropertyTest)
  test_Complicated_List(org.ognl.test.ASTPropertyTest)
  null in {true,false,null} (true)(org.ognl.test.ArithmeticAndLogicalOperatorsTest)
  null not in {true,false,null} (false)(org.ognl.test.ArithmeticAndLogicalOperatorsTest)
  5 in {true,false,null} (false)(org.ognl.test.ArithmeticAndLogicalOperatorsTest)
  5 not in {true,false,null} (true)(org.ognl.test.ArithmeticAndLogicalOperatorsTest)
  #y + "1" (11)(org.ognl.test.ArithmeticAndLogicalOperatorsTest)
  "1" + #y (11)(org.ognl.test.ArithmeticAndLogicalOperatorsTest)
  { false, true, null, 0, 1. } ([false, true, null, 0, 1.0])(org.ognl.test.ConstantTest)
  messages.format('ShowAllCount', {one}) (foo)(org.ognl.test.MethodTest)
  messages.format('ShowAllCount', {one, two}) (foo)(org.ognl.test.MethodTest)
  {theInt + 1}(org.ognl.test.PropertyArithmeticAndLogicalOperatorsTest)

Tests run: 730, Failures: 1, Errors: 15, Skipped: 0
````
In the logs we see a lot of messages like:
```
Running org.ognl.test.ArrayElementsTest
Caught exception java.lang.RuntimeException: java.io.IOException: invalid constant type: 18
Caught exception java.lang.RuntimeException: java.io.IOException: invalid constant type: 18
Tests run: 11, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.026 sec <<< FAILURE!
```
The invalid constant type exception referes to Java 8 classes. Okay, OGNL uses javassist to access the java objects. The utilized version is 3.11, which dosn't support Java 8. 
Updating javassist 3.11.0.GA to 3.20.0-GA changes the error message:

```
Results :

Failed tests:
  messages.format('ShowAllCount', one) (foo)(org.ognl.test.MethodTest)
  messages.format('ShowAllCount', {one}) (foo)(org.ognl.test.MethodTest)
  messages.format('ShowAllCount', {one, two}) (foo)(org.ognl.test.MethodTest)

Tests run: 730, Failures: 3, Errors: 0, Skipped: 0
```
When using java 1.7 all tests pass. Amazing ;)

So I did some further investigations and this are my findings:
For Java 1.8 the message.format calls resulted in those function calls:
```
messages.format('ShowAllCount', one) -> calls format(String key, Object param1)
messages.format('ShowAllCount', {one}) -> calls format(String key, Object param1)
	param1 is Array[1]... Okay, Array matches to object...
messages.format('ShowAllCount', {one, two}) -> calls format(String key, Object param1)
	param1 is Array[1, 2]... Okay, Array matches to object...
```
Then I found out following facts;
* The unit-tests pre-compile the expressions. (Hint to me: We get different results if expressions are compiled vs. just parsed - needs further investigations)
* The compiler produces different code during *ASTChain.toGetSource()*: 
 * Java 1.7: `.getMessages()).format("ShowAllCount", (java.lang.Object[]) ref1($$))`
 * Java 1.8: `.getMessages()).format("ShowAllCount",  ($w) ((org.ognl.test.objects.Simple)$2)..getOne())`
* Further investigations brought up that *OgnlRuntime.getReadMethod()* returns different methods:
 * Java 1.7: `public java.lang.String org.ognl.test.objects.Messages.format(java.lang.String,java.lang.Object[])`
 * Java 1.8: `public java.lang.String org.ognl.test.objects.Messages.format(java.lang.String,java.lang.Object)`

Okay, the root cause is: *OgnlRuntime.getReadMethod()* obtains a list of methods by calling
```
            BeanInfo info = Introspector.getBeanInfo(target);
            MethodDescriptor[] methods = info.getMethodDescriptors();
```
and iterating through this list, returning the first method that could propably match. And *info.getMethodDescriptors()* returns the list in a different order depending on the java version.

Okay, how to solve this?

My only idea would be to extend this getReadMethod() with following functionallity:
* collect all possible matchin methods
* if we have more then one candidate: 
 * calculate a 'weight' for each method depending how 'good' the arguments match
 * sort the list by wight and return the best candidate

Any other ideas?

**Comment by marvkis** [comment]
I did a short test to see how the matching for java itself behaves:
```java
    public static void main(String[] args) {
    	Messages messages = new Messages(new HashMap<Object, Object>());
        System.out.println("messages.format('ShowAllCount', one) => \"foo\" == "+messages.format("ShowAllCount", 1 )); 
        System.out.println("messages.format('ShowAllCount', {one}) => \"foo\" == "+messages.format("ShowAllCount", new Object[] { 1 } )); 
        System.out.println("messages.format('ShowAllCount', {one, two}) => \"foo\" == "+messages.format("ShowAllCount", new Object[] { 1, 2 } )); 
        System.out.println("messages.format('ShowAllCount', one, two) => \"foo\" == "+messages.format("ShowAllCount", 1, 2 )); 
    }
```
The result is:
```
messages.format('ShowAllCount', one) => "foo" == first
messages.format('ShowAllCount', {one}) => "foo" == foo
messages.format('ShowAllCount', {one, two}) => "foo" == foo
messages.format('ShowAllCount', one, two) => "foo" == haha
```

And this is the same for both java versions (1.7 & 1.8). So this would be the ideal result ;)

**Comment by lukaszlenart** [comment]
Looks like the same issue as here https://issues.apache.org/jira/browse/OGNL-250

**Comment by marvkis** [comment]
Hi @lukaszlenart ,
Thank you for this link. I havn't been aware that the OGNL project on apache is still alive ;)

Never the less I started digging a bit deeper into it (also in conjunction with #17 ) and it seems that there are two seperate implementations for finding out the appropiate method:
*OgnlRuntime.getReadMethod()* used by the compiler and *OgnlRuntime.getAppropriateMethod()* in other cases.

Might it make sense to centralize it to one implementation?

I also have seen that methods are actually called during the *compile()*. Are you aware of this? Dosn't it cause a lot of confusion in the backend?


And should wie transfer my tickets to Apache?

**Comment by lukaszlenart** [comment]
Yes, I mean, soon I want to switch all the efforts to Apache Commons OGNL and start preparing a new 4.0 release. I don't want to spend more time on extending/improving 3.x line - it's here just to keep support, fix bugs and so on.

**Comment by marvkis** [comment]
So does it make sense to fix this regressin in 3.x or shall i just fix my issue in #17 and leave this as Task for OGNL 4?

**Comment by lukaszlenart** [comment]
Up to you but OGNL4 can differ in this area so maybe there is no such issue there.

**Comment by lukaszlenart** [comment]
one remark: releasing OGNL3 is under my command, with OGNL4 won't that easy ;-)

**Comment by marvkis** [comment]
Hi,

I started digging into that method matching and implemented an initial version.
It works with all current unit tests and also fixes #17.

Please have a look on the code and decide what to do. It could fix a lot things but also be dangerous... 


And just some ideas i had while i was working in the depth of OGNL - could be interesting for OGNL4:

It might be better to have all the Matching and Conversion stuff done on "Java-Class" basis instead of Object basis. I think this is why methods are called during OGNL.compile() - to get the returning Object. When the method is not callable or failing the compiler starts struggling.
My Opinion is that the Object is not really required. I think looking taking Method.getReturnType() and working with the class would be better provides better matching capabilties - we even know what it is if the method fails or returns null.

And as further use case - it might be passible to implement autocompletion / code assist functionality based on this for OGNL ;)

Are there already some discissions on this?
